### ISSUE TITLE
Unnecessary padding and flipped padding dims in swin-attention.py

### ISSUE THREAD

**Issue by jakecole1** [cambridge-mlg/gridded-tnp#3]
Hi there,

Firstly, it looks like the padding in swin attention is being applied in reverse order via torch.nn.functional.pad. For ConstantPad2d, the tuple returned by get_padding_tuple() seems like it needs to be reversed

For example: 

```python
grid_shape = torch.tensor([91, 120])
window_sizes = torch.tensor([3, 4])

padding = (
    (3- 91 % 3) % 3,   # lat: 2
    (4 - 120% 4) % 4,  # long: 0
)
-> padding = (2, 0)
padding_tuple = (
    (2 // 2, 2 - 2// 2),   # (1, 1)
    (0 // 2, 0 - 0 // 2),   # (0, 0)
)
-> padding_tuple = ((1, 1), (0, 0))
```

So padding is being applied to the long instead of the lat as
(pad_long_left=1, pad_long_right=1, pad_lat_left=0, pad_lat_right=0)

You can recreate this issue by setting SOURCE_GRID_SIZE = [4, 90, 120], and WINDOW_SIZES = [4, 3, 3].
The padding [4, 3, 3] is being added via nn.functional.pad in the opposite dimension order as [3, 3, 4], returning the following:
```python
x shape before add_padding: torch.Size([8, 4, 90, 120, 128])
x shape after add_padding: torch.Size([8, 7, 93, 124, 128])
```
### Proposed Fix

In tnp/networks/swin-attention/py
```python
padding_tuple = tuple(itertools.chain(*padding_tuple))
``` 
Which can be fixed by 

```python
 padding_tuple = tuple(itertools.chain(*reversed(padding_tuple)))
``` 

### Additional issue

Another issue that might require some additional thought is the way padding is currently computed. The current implementation pads even when a dimension is already divisible by the window size (for example, when 4 % 4 = 0):

Again setting SOURCE_GRID_SIZE = [4, 90, 120], and WINDOW_SIZES = [4, 3, 3] (and assuming the reverse logic from above has been fixed): 
```python
x shape before add_padding: torch.Size([8, 4, 90, 120, 128])
x shape after add_padding: torch.Size([8, 8, 93, 123, 128])
``` 
However, x should not need any padding and should not be changed as 

```python
window_sizes[0] divides time_steps
window_sizes[1] divides grid_encoder_grid_shape[0]
window_sizes[2] divides grid_encoder_grid_shape[1
``` 

```python
padding = tuple(
window_sizes[i] - grid_shape[i] % window_sizes[i]
for i in range(len(grid_shape))
)
``` 

### Proposed fix
A simple fix is to apply another modulo operation so that no padding is added when the dimension is already divisible by the window size:

```python
padding = tuple(
(window_sizes[i] - grid_shape[i] % window_sizes[i]) % window_sizes[i]
for i in range(len(grid_shape))
)
``` 

This also matches the pytorch implementation: [](https://docs.pytorch.org/vision/main/_modules/torchvision/models/swin_transformer.html)

Thanks in advance for taking a look!
